### PR TITLE
[render-test] Add manifest parser for render-test-runner

### PR DESCRIPTION
**Pull request by zmiao**
Currently our render-test-runner contains several hardcoded file paths, this introduces more difficulties of potability and raises the complexity and ambiguity of CLI argument parsing. It also decreases the flexibility of further development. Due to these reasons, we decide to introduce platform dependent manifest file that will contain all of the necessary resources that we need for running the render-test-runner, we can get rid of all the hardcoded files afterwards.

### CODE REVIEW

File: render-test/manifest_parser.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include "filesystem.hpp"
+
+class ManifestParser {
+public:
+    static ManifestParser& getInstance() {
+        static ManifestParser instance;
```
Comment (pozdnyakov) [inline render-test/manifest_parser.hpp]: IMO we could do without a singleton, see the comments below

File: render-test/manifest_parser.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        return instance;
+    }
+    const std::string getVendorPath() const;
+    const std::string getIntegrationPath() const;
+    const mbgl::filesystem::path getBaseTestPath() const;
```
Comment (pozdnyakov) [inline render-test/manifest_parser.hpp]: no need to return a const type by value, could be `const mbgl::filesystem::path&`

File: render-test/manifest_parser.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#include <vector>
+#include "filesystem.hpp"
+
+class ManifestParser {
+public:
+    static ManifestParser& getInstance() {
+        static ManifestParser instance;
+        return instance;
+    }
+    const std::string getVendorPath() const;
```
[the remark this answers is not in the document]
Comment (zmiao) [reply render-test/manifest_parser.hpp]: Updated.

File: render-test/manifest_parser.cpp [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    return mbgl::filesystem::path{};
+}
+
+} // namespace
+
+mbgl::optional<Manifest> ManifestParser::parseManifest(const std::string& rootPath,
+                                                       const std::vector<std::string>& testNames,
+                                                       const std::string& testFilter) {
+    static Manifest manifest;
+    auto filePath = mbgl::filesystem::path(rootPath);
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: Let's pass the `manifestPath` from the cmd line

Comment (pozdnyakov) [reply render-test/manifest_parser.cpp]: and it looks like we do not need `rootPath` in that case

File: render-test/manifest_parser.cpp [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        const auto& assetPathValue = document["asset_path"];
+        if (!assetPathValue.IsString()) {
+            mbgl::Log::Warning(
+                mbgl::Event::General, "Invalid assetPath is provoided inside the manifest file: %s", filePath.c_str());
+            return {};
+        }
+
+        manifest.assetPath = getValidPath(rootPath, assetPathValue.GetString()).string();
+        if (manifest.assetPath.back() != '/') {
+            manifest.assetPath.push_back('/');
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: why is that needed? could it be done with `mbgl::filesystem` api?

File: render-test/manifest_parser.cpp (hunk 4 above)
Comment (zmiao) [reply render-test/manifest_parser.cpp]: `rootPath` is necessary in case the paths inside manifest are relative.

File: render-test/runner.hpp [hunk 6]
```diff
@@ -32,5 +30,5 @@ class TestRunner {
         mbgl::Map map;
     };
     std::unordered_map<std::string, std::unique_ptr<Impl>> maps;
-    std::string testRootPath{TEST_RUNNER_ROOT_PATH};
+    const Manifest& manifest;
```
Comment (pozdnyakov) [inline render-test/runner.hpp]: IMO `TestRunner` should own `manifest`

File: render-test/manifest_parser.cpp [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    return url;
+}
+
+std::string prependFileScheme(const std::string& url) {
+    static const std::string fileScheme("file://");
+    return fileScheme + url;
+}
+} // namespace
+
+Manifest::Manifest() : manifestPath(), testRootPath(), vendorPath(), assetPath(), ignores(), testPaths() {}
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: `manifestPath()` - no need, 
`Manifest::Manifest() = default;`
`Manifest::~Manifest() = default;`

File: render-test/manifest_parser.cpp [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    mbgl::Log::Warning(mbgl::Event::General, "Invalid path is provoided inside the manifest file: %s", path.c_str());
+    return mbgl::filesystem::path{};
+}
+
+} // namespace
+
+mbgl::optional<Manifest> ManifestParser::parseManifest(const std::string& manifestPath,
+                                                       const std::vector<std::string>& testNames,
+                                                       const std::string& testFilter) {
+    static Manifest manifest;
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: why is it static?^

File: render-test/manifest_parser.cpp [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+    const auto filePath = mbgl::filesystem::path(manifestPath);
+    manifest.manifestPath = manifestPath.substr(0, manifestPath.find(filePath.filename()));
+
+    auto contents = readJson(filePath);
+    if (!contents.is<mbgl::JSDocument>()) {
+        mbgl::Log::Error(mbgl::Event::General, "Provided manifest file: %s is not a valid json", filePath.c_str());
+        return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: `nullopt`

File: render-test/manifest_parser.cpp [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+        const auto& expectationPathValue = document["expectation_paths"];
+        if (!expectationPathValue.IsArray()) {
+            mbgl::Log::Warning(mbgl::Event::General,
+                               "Invalid expectationPath is provoided inside the manifest file: %s",
+                               filePath.c_str());
+            return {};
+        }
+        for (const auto& value : expectationPathValue.GetArray()) {
+            if (!value.IsString()) {
+                return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: mbgl::Log::Warning here?

File: render-test/manifest_parser.cpp [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+            }
+            expectationPaths.emplace_back(getValidPath(manifest.manifestPath, value.GetString()));
+            if (expectationPaths.back().empty()) {
+                return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: ditto

File: render-test/manifest_parser.cpp [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+    if (document.HasMember("ignore_paths")) {
+        const auto& ignorePathValue = document["ignore_paths"];
+        if (!ignorePathValue.IsArray()) {
+            mbgl::Log::Warning(
+                mbgl::Event::General, "Invalid ignorePath is provoided inside the manifest file: %s", filePath.c_str());
+            return {};
+        }
+        for (const auto& value : ignorePathValue.GetArray()) {
+            if (!value.IsString()) {
+                return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: ditto

File: render-test/manifest_parser.cpp [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+            }
+            ignorePaths.emplace_back(getValidPath(manifest.manifestPath, value.GetString()));
+            if (ignorePaths.back().empty()) {
+                return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: ditto and `nullopt` too

File: render-test/manifest_parser.cpp [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+            }
+
+            if (testPath.path().filename() == "style.json") {
+                testPaths.emplace_back(testPath, getTestExpectations(testPath, path, expectationPaths));
+            }
+        }
+    }
+
+    manifest.testRootPath = enbaleProbeTest ? probeTestPath.string() : baseTestPath.string();
+    if (manifest.testRootPath.back() == '/') {
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: is this condition really needed? I'd assume `mbgl::filesystem` will return string in the same way all the time :thinking:

File: render-test/runner.hpp [hunk 15]
```diff
@@ -11,11 +14,14 @@ struct TestMetadata;
 
 class TestRunner {
 public:
-    TestRunner() = default;
-    explicit TestRunner(const std::string& testRootPath);
+    explicit TestRunner(mbgl::optional<Manifest> manifest_);
```
Comment (pozdnyakov) [inline render-test/runner.hpp]: explicit TestRunner(Manifest manifest_); // we cannot have TestRunner without a manifest

File: render-test/manifest_parser.cpp [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+    } else {
+        for (const auto& id : testNames) {
+            paths.emplace_back(baseTestPath / id);
+        }
+        if (paths.empty()) {
+            paths.emplace_back(baseTestPath);
+        }
+    }
+
+    // Recursively traverse through the test paths and collect test dirπectories containing "style.son".
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: dirπectories -> directories

Comment (pozdnyakov) [reply render-test/manifest_parser.cpp]: style.son -> style.json

File: render-test/manifest_parser.cpp [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+            if (ignorePaths.back().empty()) {
+                return {};
+            }
+        }
+        manifest.ignores = parseIgnores(ignorePaths);
+    }
+
+    bool enbaleProbeTest{false};
+    std::vector<mbgl::filesystem::path> paths;
+    // Seperate Probe tests and normal tests, since they have different testRoot
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: Could probe tests have a separate manifest file instead?

File: render-test/manifest_parser.cpp [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+    mbgl::filesystem::path file = std::regex_replace(url, regex, vendorPath);
+    if (mbgl::filesystem::exists(file.parent_path())) {
+        return removeURLArguments(file.string());
+    }
+
+    if (glyphsPath && mbgl::filesystem::exists(file.parent_path().parent_path())) {
+        return removeURLArguments(file.string());
+    }
+
+    return {};
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: nullopt (and in other places too)

File: render-test/manifest_parser.cpp [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+                .value_or(localizeLocalURL(spriteValue.GetString()).value_or(spriteValue.GetString())));
+        spriteValue.Set<std::string>(path, document.GetAllocator());
+    }
+}
+
+mbgl::optional<std::string> Manifest::localizeLocalURL(const std::string& url, bool glyphsPath) const {
+    static const std::regex regex{"local://"};
+    if (auto vendorPath = getVendorPath(url, regex, glyphsPath)) {
+        return vendorPath;
+    } else {
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: nit: else is not needed

File: render-test/manifest_parser.cpp [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+        return getIntegrationPath(url, "", regex, glyphsPath);
+    }
+}
+
+mbgl::optional<std::string> Manifest::localizeHttpURL(const std::string& url) const {
+    static const std::regex regex{"http://localhost:2900"};
+    if (auto vendorPath = getVendorPath(url, regex)) {
+        return vendorPath;
+    } else {
```
Comment (pozdnyakov) [inline render-test/manifest_parser.cpp]: ditto

File: render-test/manifest_parser.cpp (hunk 14 above)
Comment (zmiao) [reply render-test/manifest_parser.cpp]: Nope, it will return the original path string, which may contain "/" or not.

File: render-test/manifest_parser.cpp (hunk 13 above)
Comment (zmiao) [reply render-test/manifest_parser.cpp]: Here log is not necessary, there is a log inside `getValidPath` in case empty path is returned.

File: render-test/manifest_parser.cpp (hunk 17 above)
Comment (zmiao) [reply render-test/manifest_parser.cpp]: I will add a separate one.
